File: typescript/deployments/gcp-cloud-function/src/dao/firestore-person-dao-impl.ts
import {PersonDao} from "../../../../core/src/dao/person-dao";
import {Optional} from "typescript-optional";
import {Person} from "../../../../core/src/model/person";
import {Firestore} from "@google-cloud/firestore";
import {v4 as uuidv4} from "uuid";


export class FirestorePersonDaoImpl extends PersonDao {

    private readonly table: string = 'person';

    private readonly firestoreDatabase: Firestore;

    constructor() {
        super();
        this.firestoreDatabase = new Firestore();
    }

    public async getById(id: string): Promise<Optional<Person>> {
        console.log(`Getting person by id: ${id}`);

        const document: FirebaseFirestore.DocumentSnapshot<FirebaseFirestore.DocumentData> = await this.firestoreDatabase
            .collection(this.table)
            .doc(id)
            .get();

        console.log(document);

        if (document.exists) {
            return Optional.of(
                new Person(
                    document.id,
                    document.get('firstName'),
                    document.get('lastName'),
                    new Date(document.get('dateOfBirth'))
                )
            )
        }

        return Optional.empty();
    }

    public async save(person: Person): Promise<void> {
        console.log('Saving person')

        // Give the person an ID if it doesn't have one.
        if (! person.id) {
            person.id = uuidv4().toString()
        }

        await this.firestoreDatabase
            .collection(this.table)
            .doc(person.id) // Unique ID/primary key of document/object.
            .set({
                "dateOfBirth": person.dateOfBirth,
                "firstName": person.firstName,
                "lastName": person.lastName
            })
            .then(documentReference => {
                console.log(`Document write time: ${documentReference.writeTime}`);
            });
    }

}

File: dotnet/deployments/gcp-cloud-function/dao/FirestorePersonDaoImpl.cs
using core.dao;
using core.model;
using Google.Cloud.Firestore;
using Newtonsoft.Json;

namespace gcp_cloud_function;

public class FirestorePersonDaoImpl : PersonDao
{
    private readonly FirestoreDb _firestoreDb = FirestoreDb.Create();

    private static readonly string PERSON_TABLE_NAME = "person";

    private static readonly string ID_FIELD_NAME = "id";

    public async Task<Person?> GetById(Guid id)
    {
        Console.WriteLine("Getting person by id {0}", id);
        
        var snapshot = await _firestoreDb.Collection(PERSON_TABLE_NAME)
            .Document(id.ToString())
            .GetSnapshotAsync();

        Console.WriteLine("snapshot: {0}", snapshot.Exists);
        if (snapshot.Exists)
        {
            return new Person(
                Guid.Parse(snapshot.Id),
                snapshot.GetValue<string>("firstName"),
                snapshot.GetValue<string>("lastName"),
                 DateTimeOffset.FromUnixTimeSeconds(snapshot.GetValue<int>("dateOfBirth"))
            );
        }

        return null;
    }

    public async Task<bool> Save(Person person)
    {
        Console.WriteLine("Saving person {0}", person);

        var personObj = new
        {
            firstName = person.FirstName,
            lastName = person.LastName,
            dateOfBirth = person.DateOfBirth?.ToUnixTimeSeconds()
        };
        
        Console.WriteLine("personObj is {0}", personObj);
        
        await _firestoreDb.Collection(PERSON_TABLE_NAME)
            .Document(person.id.ToString())
            .SetAsync(personObj);
        
        return true;
    }
}